frontend/catalyst/compilation_pipelines.py

```diff
@@ -550,9 +546,111 @@ def __call__(self, *args, **kwargs):
+        return jax.pure_callback(qfunc, qfunc.jaxpr.out_avals, *args, vectorized=False, **kwargs)
+
+    def get_derivative_qfunc(self, argnums):
+        """Compile a function computing the derivative of the wrapped QJIT for the given argnums."""
+
+        argnum_key = "".join(str(idx) for idx in argnums)
+        if argnum_key in self.deriv_qfuncs:
+            return self.deriv_qfuncs[argnum_key]
+
+        # Here we define the signature for the new QJIT object explicitly, rather than relying on
+        # functools.wrap, in order to guarantee compilation is triggered on instantiation.
+        # The signature of the original QJIT object is guaranteed to be defined by now, located
+        # in QJIT.c_sig, however we don't update the original function with these annotations.
+        annotations = {}
+        updated_params = []
+        signature = inspect.signature(self.qfunc)
+        for idx, (arg_name, param) in enumerate(signature.parameters.items()):
+            annotations[arg_name] = self.qfunc.c_sig[idx]
+            updated_params.append(param.replace(annotation=annotations[arg_name]))
+
+        def deriv_wrapper(*args, **kwargs):
+            return catalyst.grad(self.qfunc, argnum=argnums)(*args, **kwargs)
+
+        deriv_wrapper.__name__ = "deriv_" + self.qfunc.__name__
+        deriv_wrapper.__annotations__ = annotations
+        deriv_wrapper.__signature__ = signature.replace(parameters=updated_params)
+
+        self.deriv_qfuncs[argnum_key] = QJIT(deriv_wrapper, self.qfunc.compile_options)
+        return self.deriv_qfuncs[argnum_key]
+
+    def compute_jvp(self, primals, tangents):
+        """Compute the set of results and JVPs for a QJIT function."""
+
+        # Optimization: Do not compute Jacobians for arguments which do not participate in
+        #               differentiation.
+        # TODO: replace with symbolic zero support
+        argnums = []
+        for idx, tangent in enumerate(tangents):
+            if qml.math.is_abstract(tangent, like="jax") or jnp.any(jnp.atleast_1d(tangent) != 0.0):
+                argnums.append(idx)
+
+        results = self.wrap_callback(self.qfunc, *primals)
+        derivatives = self.wrap_callback(self.get_derivative_qfunc(argnums), *primals)
+
+        jvps = [jnp.zeros_like(results[res_idx]) for res_idx in range(len(results))]
+        for arg_idx, tangent in enumerate(tangents):
+            # Skip JVP for arguments which do not participate in differentiation.
+            if arg_idx not in argnums:
+                continue
+            for res_idx in range(len(results)):
+                deriv_idx = arg_idx * len(results) + res_idx
+                num_axes = 0 if tangent.ndim == 0 else 1
+                jvp = jnp.tensordot(jnp.transpose(derivatives[deriv_idx]), tangent, axes=num_axes)
+                jvps[res_idx] = jvps[res_idx] + jvp
+
+        if len(results) == 1:
+            results = results[0]
+        if len(jvps) == 1:
+            jvps = jvps[0]
```

Comment:
Yeah I think if you consistently used JAX's PyTree infrastructure with the flattening and unflattening functions this wouldn't be an issue, so this could definitely be improved.